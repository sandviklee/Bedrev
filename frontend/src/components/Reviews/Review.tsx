import { FiUser, FiStar } from "react-icons/fi";
import { ReviewType } from "../../types/types";
import { gql, useMutation, useQuery } from "@apollo/client";
import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import Icon from "../Icon/Icon";
import { ReviewJSON } from "../../types/types";

const Review = ({ title, rating, description, id }: ReviewType) => {
  const authToken = localStorage.getItem("auth-token"); //Global state management.
  const [reviewId, setReviewId] = useState<number | null>(null);
  const navigate = useNavigate();

  /**
   * Query for getting all the reviews
   * the user has.
   */
  const USER_QUERY = gql`
    query {
      me {
        reviews {
          id
        }
      }
    }
  `;

  const DELETE_MUTATION = gql`
        mutation {
            deleteReview(id: ${reviewId})
        }
    `;

  const [deleteReview] = useMutation(DELETE_MUTATION, {
    onCompleted: () => {
      navigate(0);
    },
    context: { headers: { authorization: `Bearer ${authToken}` } },
  });

  const { data } = useQuery(USER_QUERY, {
    fetchPolicy: "cache-first", // Used for first execution
    nextFetchPolicy: "cache-first", // Used for subsequent executions
    context: { headers: { authorization: `Bearer ${authToken}` } },
  });

  /**
   * Use effect for checking if the review
   * is made by the authorized user
   * and if so, sets the review id.
   */
  useEffect(() => {
    setReviewId(null);
    if (data !== undefined) {
      data["me"]["reviews"].map((review: ReviewJSON) => {
        if (review["id"] == id) {
          setReviewId(review.id);
        }
      });
    }
  }, [data, title]);

  return (
    <>
      <div
        style={{ backgroundColor: reviewId ? "#C4A8F4" : "#E2DAEB" }}
        className="px-[20px] py-[20px] flex-row sm:h-fit max-h-fit min-h-[200px] w-[350px] rounded-md font-whyte"
      >
        <div className="flex">
          <p className="flex w-full justify-start leading-[40px] text-[90%] font-medium ">
            {title}
          </p>
          {reviewId && (
            <button onClick={() => deleteReview()} className="flex justify-end">
              <Icon icon="Trash" />
            </button>
          )}
        </div>

        <div className="flex items-center align-middle gap-1 py-2">
          <div className="text-2xl pb-2">
            <FiStar />
          </div>
          <p className="text-lg font-medium">{rating}/5</p>
        </div>
        <p className="leading-[30px] font-regular text-[16px] sm:h-[120px] max-h-[180px] min-h-[160px]">
          {description}
        </p>
        <p className="flex gap-1 py-2 justify-end text-base font-medium items-center ">
          <div className="text-2xl pb-1">
            <FiUser />
          </div>
          <p>Anonym</p>
        </p>
      </div>
    </>
  );
};

export default Review;
