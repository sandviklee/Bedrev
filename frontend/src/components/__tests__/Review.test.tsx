import { MockedProvider } from "@apollo/client/testing";
import { describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import Review from "../Reviews/Review.tsx";
import { BrowserRouter as Router } from "react-router-dom";
import { gql } from "@apollo/client";

describe("Review component", () => {
  const mockReview = {
    title: "Test",
    rating: 5,
    description: "This is a test",
    id: 1,
  };

  const mocks = [
    {
      request: {
        query: gql`
          query {
            me {
              reviews {
                id
              }
            }
          }
        `,
      },
      result: {
        data: {
          me: {
            reviews: [
              { id: 1 },
              // Add more reviews if needed
            ],
          },
        },
      },
    },
    {
      request: {
        query: gql`
          mutation {
            deleteReview(id: 1)
          }
        `,
      },
      result: {
        data: {
          deleteReview: true, // Assume the mutation returns a boolean
        },
      },
    },
  ];

  test("renders correctly", () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Router>
          <Review {...mockReview} />
        </Router>
      </MockedProvider>,
    );

    expect(screen.getByText("Test")).toBeInTheDocument();
    expect(screen.getByText("This is a test")).toBeInTheDocument();
    expect(screen.getByText("5/5")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Router>
          <Review {...mockReview} />
        </Router>
      </MockedProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
