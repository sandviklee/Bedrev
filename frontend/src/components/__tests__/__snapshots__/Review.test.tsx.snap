// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Review component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-[20px] py-[20px] flex-row sm:h-fit max-h-fit min-h-[200px] w-[350px] rounded-md font-whyte"
    style="background-color: rgb(226, 218, 235);"
  >
    <div
      class="flex"
    >
      <p
        class="flex w-full justify-start leading-[40px] text-[90%] font-medium "
      >
        Test
      </p>
    </div>
    <div
      class="flex items-center align-middle gap-1 py-2"
    >
      <div
        class="text-2xl pb-2"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
          />
        </svg>
      </div>
      <p
        class="text-lg font-medium"
      >
        5/5
      </p>
    </div>
    <p
      class="leading-[30px] font-regular text-[16px] sm:h-[120px] max-h-[180px] min-h-[160px]"
    >
      This is a test
    </p>
    <p
      class="flex gap-1 py-2 justify-end text-base font-medium items-center "
    />
    <div
      class="text-2xl pb-1"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
        />
        <circle
          cx="12"
          cy="7"
          r="4"
        />
      </svg>
    </div>
    <p>
      Anonym
    </p>
    <p />
  </div>
</DocumentFragment>
`;
